src/lib/krb5/krb/gc_via_tkt.c

```diff
@@ -253,18 +281,23 @@ krb5int_process_tgs_reply(krb5_context context,
     /* make sure the response hasn't been tampered with..... */
     retval = 0;
 
-    if (s4u2self && !IS_TGS_PRINC(dec_rep->ticket->server)) {
-        /* Final hop, check whether KDC supports S4U2Self */
-        if (krb5_principal_compare(context, dec_rep->client, in_cred->server))
+    if (s4u && !IS_TGS_PRINC(dec_rep->ticket->server)) {
+        /* Final hop, check whether KDC supports S4U */
+        if (krb5_principal_compare(context, dec_rep->client, tkt->client))
```

Comment:
Applying this test to S4U2Proxy requests seems like an ancillary behavior change.  Can you comment on that?  If it's a non-trivial bugfix we should make it a separate commit with a ticket number.  If it's a small thing we should mention it in the commit message.

Reply:
Per MS-SFU:
`Services can detect whether the KDC supports S4U by checking the cname of the returned ticket.
KDCs that do not support S4U ignore the S4U2self and S4U2proxy data and return a service ticket
with the cname containing the name of the service that made the request ([RFC4120] section 3.3.3).
In service tickets from KDCs that support S4U, the cname contains the name of the user.`

This is repeated three times, in the introduction 1.5, for S4U2Self 3.1.5.1.2, and for S4U2proxy 3.1.5.2.4.
So I think this change is right, but maybe I got carried away by wanting to unify the s4u logic. I'll rework this commit and either avoid the unrelated changes or move it to its own commit.

Reply:
I think the change is fine (assuming we aren't worried about the degenerate case where the evidence ticket is for the intermediate server itself).  I just stumbled over it since it didn't seem to be related to the purpose of this commit.

I'm also a little surprised that Luke's code didn't check for this somewhere.  If the KDC doesn't support S4U2Proxy, do we actually attempt continue on with a ticket from the intermediate service instead of the intended client?

Reply:
> assuming we aren't worried about the degenerate case where the evidence ticket is for the intermediate server itself

and this case could still work by using alias names.

> If the KDC doesn't support S4U2Proxy, do we actually attempt continue on with a ticket from the intermediate service instead of the intended client?

I think in MIT the code we generally make a better check in the caller, to verify that the return principal is the same as the one we expect, but I think is worth having the basic check that we can do within krb5_get_credentials() where we don't know the expected client name.
gss: https://github.com/krb5/krb5/blob/09c9b7d6f64767429e90ad11a529e6ffa9538043/src/lib/gssapi/krb5/init_sec_context.c#L206
krb5: https://github.com/krb5/krb5/blob/ebbc6e8e99ee9d5d757411200a6a3173171774df/src/lib/krb5/krb/s4u_creds.c#L816


I used to think that check against the expected principal was required, since I wasn't sure if the 2nd ticket is protected or it could be tempered with, but following the discussion in https://github.com/heimdal/heimdal/pull/595  I ran some tests by changing 2nd ticket on wire, and have realized it was actually protected as part of request-body, see: https://github.com/heimdal/heimdal/blob/a7bb4504f2f3dfb276f8aa154858f46ed1063011/kdc/krb5tgs.c#L1093
And if the request is protected, then so is client name inside encrypted reply and therefore the unencrypted client name too, as they must match.